### ISSUE TITLE
Freeze after twice POST "/oauth/token" requests with same credentials

### ISSUE THREAD

**Issue by ViTVetal** [doorkeeper-gem/doorkeeper#1133]
I have noticed this behavior several times. 
1) start of Rails app
2) Two  POST "/oauth/token" requests with same credentials (at the same time, almost without interruption)
3) 

> rails s
> => Booting Puma
> => Rails 5.1.6 application starting in development 
> => Run `rails server -h` for more startup options
> Puma starting in single mode...
> * Version 3.11.4 (ruby 2.5.0-p0), codename: Love Song
> * Min threads: 5, max threads: 5
> * Environment: development
> * Listening on tcp://0.0.0.0:3000
> Use Ctrl-C to stop
> Started POST "/oauth/token" for 192.168.2.102 at 2018-08-13 16:24:16 +0300
> Started POST "/oauth/token" for 192.168.2.102 at 2018-08-13 16:24:16 +0300
> Cannot render console from 192.168.2.102! Allowed networks: 127.0.0.1, ::1, 127.0.0.0/127.255.255.255
> Cannot render console from 192.168.2.102! Allowed networks: 127.0.0.1, ::1, 127.0.0.0/127.255.255.255

4) freeze of app. I can restart it only with `kill -9 $(lsof -i tcp:3000 -t)`

Migration:

```ruby
class CreateDoorkeeperTables < ActiveRecord::Migration[5.1]
  def change
    create_table :oauth_access_tokens do |t|
      t.integer  :resource_owner_id
      t.integer  :application_id
      t.string   :token,                  null: false
      t.string   :refresh_token
      t.integer  :expires_in
      t.datetime :revoked_at
      t.datetime :created_at,             null: false
      t.string   :scopes
    end

    add_index :oauth_access_tokens, :token, unique: true
    add_index :oauth_access_tokens, :resource_owner_id
    add_index :oauth_access_tokens, :refresh_token, unique: true
    add_foreign_key(
      :oauth_access_tokens,
      :passenger_users,
      column: :resource_owner_id
    )
  end
end
```
initializers/doorkeeper.rb

```ruby
Doorkeeper.configure do
  # Change the ORM that doorkeeper will use (needs plugins)
  orm :active_record

  # This block will be called to check whether the resource owner is authenticated or not.
  resource_owner_authenticator do
    fail "Please configure doorkeeper resource_owner_authenticator block located in #{__FILE__}"
    # Put your resource owner authentication logic here.
    # Example implementation:
    #   User.find_by_id(session[:user_id]) || redirect_to(new_user_session_url)
  end
  #Make by phone instead email
  resource_owner_from_credentials do |_routes|
    if params[:scope].present?
      case params[:scope]
        when "passenger"
          PassengerUser.authenticate(params[:email], params[:password])
        when "driver"
          DriverUser.authenticate(params[:email], params[:password])
      end
    else 
      PassengerUser.authenticate(params[:email], params[:password])
    end
  end

  grant_flows %w(password)

  skip_authorization do
    true
  end

  # If you want to restrict access to the web interface for adding oauth authorized applications, you need to declare the block below.
  # admin_authenticator do
  #   # Put your admin authentication logic here.
  #   # Example implementation:
  #   Admin.find_by_id(session[:admin_id]) || redirect_to(new_admin_session_url)
  # end

  # Authorization Code expiration time (default 10 minutes).
  # authorization_code_expires_in 10.minutes

  # Access token expiration time (default 2 hours).
  # If you want to disable expiration, set this to nil.
  # access_token_expires_in 2.hours

  # Assign a custom TTL for implicit grants.
  # custom_access_token_expires_in do |oauth_client|
  #   oauth_client.application.additional_settings.implicit_oauth_expiration
  # end

  # Use a custom class for generating the access token.
  # https://github.com/doorkeeper-gem/doorkeeper#custom-access-token-generator
  # access_token_generator '::Doorkeeper::JWT'

  # The controller Doorkeeper::ApplicationController inherits from.
  # Defaults to ActionController::Base.
  # https://github.com/doorkeeper-gem/doorkeeper#custom-base-controller
  # base_controller 'ApplicationController'

  # Reuse access token for the same resource owner within an application (disabled by default)
  # Rationale: https://github.com/doorkeeper-gem/doorkeeper/issues/383
  #reuse_access_token

  # Issue access tokens with refresh token (disabled by default)
  use_refresh_token

  # Provide support for an owner to be assigned to each registered application (disabled by default)
  # Optional parameter confirmation: true (default false) if you want to enforce ownership of
  # a registered application
  # Note: you must also run the rails g doorkeeper:application_owner generator to provide the necessary support
  # enable_application_owner confirmation: false

  # Define access token scopes for your provider
  # For more information go to
  # https://github.com/doorkeeper-gem/doorkeeper/wiki/Using-Scopes
  default_scopes :passenger

  optional_scopes :driver

  # Change the way client credentials are retrieved from the request object.
  # By default it retrieves first from the `HTTP_AUTHORIZATION` header, then
  # falls back to the `:client_id` and `:client_secret` params from the `params` object.
  # Check out https://github.com/doorkeeper-gem/doorkeeper/wiki/Changing-how-clients-are-authenticated
  # for more information on customization
  # client_credentials :from_basic, :from_params

  # Change the way access token is authenticated from the request object.
  # By default it retrieves first from the `HTTP_AUTHORIZATION` header, then
  # falls back to the `:access_token` or `:bearer_token` params from the `params` object.
  # Check out https://github.com/doorkeeper-gem/doorkeeper/wiki/Changing-how-clients-are-authenticated
  # for more information on customization
  # access_token_methods :from_bearer_authorization, :from_access_token_param, :from_bearer_param

  # Change the native redirect uri for client apps
  # When clients register with the following redirect uri, they won't be redirected to any server and the authorization code will be displayed within the provider
  # The value can be any string. Use nil to disable this feature. When disabled, clients must provide a valid URL
  # (Similar behaviour: https://developers.google.com/accounts/docs/OAuth2InstalledApp#choosingredirecturi)
  #
  # native_redirect_uri 'urn:ietf:wg:oauth:2.0:oob'

  # Forces the usage of the HTTPS protocol in non-native redirect uris (enabled
  # by default in non-development environments). OAuth2 delegates security in
  # communication to the HTTPS protocol so it is wise to keep this enabled.
  #
  # Callable objects such as proc, lambda, block or any object that responds to
  # #call can be used in order to allow conditional checks (to allow non-SSL
  # redirects to localhost for example).
  #
  # force_ssl_in_redirect_uri !Rails.env.development?
  #
  # force_ssl_in_redirect_uri { |uri| uri.host != 'localhost' }

  # Specify what redirect URI's you want to block during creation. Any redirect
  # URI is whitelisted by default.
  #
  # You can use this option in order to forbid URI's with 'javascript' scheme
  # for example.
  #
  # forbid_redirect_uri { |uri| uri.scheme.to_s.downcase == 'javascript' }

  # Specify what grant flows are enabled in array of Strings. The valid
  # strings and the flows they enable are:
  #
  # "authorization_code" => Authorization Code Grant Flow
  # "implicit"           => Implicit Grant Flow
  # "password"           => Resource Owner Password Credentials Grant Flow
  # "client_credentials" => Client Credentials Grant Flow
  #
  # If not specified, Doorkeeper enables authorization_code and
  # client_credentials.
  #
  # implicit and password grant flows have risks that you should understand
  # before enabling:
  #   http://tools.ietf.org/html/rfc6819#section-4.4.2
  #   http://tools.ietf.org/html/rfc6819#section-4.4.3
  #
  # grant_flows %w[authorization_code client_credentials]

  # Hook into the strategies' request & response life-cycle in case your
  # application needs advanced customization or logging:
  #
  # before_successful_strategy_response do |request|
  #   puts "BEFORE HOOK FIRED! #{request}"
  # end
  #
  # after_successful_strategy_response do |request, response|
  #   puts "AFTER HOOK FIRED! #{request}, #{response}"
  # end

  # Under some circumstances you might want to have applications auto-approved,
  # so that the user skips the authorization step.
  # For example if dealing with a trusted application.
  # skip_authorization do |resource_owner, client|
  #   client.superapp? or resource_owner.admin?
  # end

  # WWW-Authenticate Realm (default "Doorkeeper").
  # realm "Doorkeeper"
end

Doorkeeper.configuration.token_grant_types << "password"
```

**Comment by nbulaj** [comment]
`Doorkeeper.configuration.token_grant_types << "password"` is a freezed option, you can't modify it this way. You could configure it using `grant_flows` option like

```ruby
Doorkeeper.configure do
  # ...

  grant_flows %w[password authorization_code client_credentials]
end
```

**Comment by ViTVetal** [comment]
@nbulaj no. For a single request or few requests with some delay - all works fine. The problem occurs only when two requests performs at the same time.
I changed doorkeeper.rb (also I tried grant_flows %w[password] and grant_flows %w(password)):

```
Doorkeeper.configure do
  # Change the ORM that doorkeeper will use (needs plugins)
  orm :active_record

  # This block will be called to check whether the resource owner is authenticated or not.
  resource_owner_authenticator do
    fail "Please configure doorkeeper resource_owner_authenticator block located in #{__FILE__}"
    # Put your resource owner authentication logic here.
    # Example implementation:
    #   User.find_by_id(session[:user_id]) || redirect_to(new_user_session_url)
  end
  #Make by phone instead email
  resource_owner_from_credentials do |_routes|
    if params[:scope].present?
      case params[:scope]
        when "passenger"
          PassengerUser.authenticate(params[:email], params[:password])
        when "driver"
          DriverUser.authenticate(params[:email], params[:password])
      end
    else 
      PassengerUser.authenticate(params[:email], params[:password])
    end
  end

  grant_flows %w[password authorization_code client_credentials]

  skip_authorization do
    true
  end

  # If you want to restrict access to the web interface for adding oauth authorized applications, you need to declare the block below.
  # admin_authenticator do
  #   # Put your admin authentication logic here.
  #   # Example implementation:
  #   Admin.find_by_id(session[:admin_id]) || redirect_to(new_admin_session_url)
  # end

  # Authorization Code expiration time (default 10 minutes).
  # authorization_code_expires_in 10.minutes

  # Access token expiration time (default 2 hours).
  # If you want to disable expiration, set this to nil.
  # access_token_expires_in 2.hours

  # Assign a custom TTL for implicit grants.
  # custom_access_token_expires_in do |oauth_client|
  #   oauth_client.application.additional_settings.implicit_oauth_expiration
  # end

  # Use a custom class for generating the access token.
  # https://github.com/doorkeeper-gem/doorkeeper#custom-access-token-generator
  # access_token_generator '::Doorkeeper::JWT'

  # The controller Doorkeeper::ApplicationController inherits from.
  # Defaults to ActionController::Base.
  # https://github.com/doorkeeper-gem/doorkeeper#custom-base-controller
  # base_controller 'ApplicationController'

  # Reuse access token for the same resource owner within an application (disabled by default)
  # Rationale: https://github.com/doorkeeper-gem/doorkeeper/issues/383
  #reuse_access_token

  # Issue access tokens with refresh token (disabled by default)
  use_refresh_token

  # Provide support for an owner to be assigned to each registered application (disabled by default)
  # Optional parameter confirmation: true (default false) if you want to enforce ownership of
  # a registered application
  # Note: you must also run the rails g doorkeeper:application_owner generator to provide the necessary support
  # enable_application_owner confirmation: false

  # Define access token scopes for your provider
  # For more information go to
  # https://github.com/doorkeeper-gem/doorkeeper/wiki/Using-Scopes
  default_scopes :passenger

  optional_scopes :driver

  # Change the way client credentials are retrieved from the request object.
  # By default it retrieves first from the `HTTP_AUTHORIZATION` header, then
  # falls back to the `:client_id` and `:client_secret` params from the `params` object.
  # Check out https://github.com/doorkeeper-gem/doorkeeper/wiki/Changing-how-clients-are-authenticated
  # for more information on customization
  # client_credentials :from_basic, :from_params

  # Change the way access token is authenticated from the request object.
  # By default it retrieves first from the `HTTP_AUTHORIZATION` header, then
  # falls back to the `:access_token` or `:bearer_token` params from the `params` object.
  # Check out https://github.com/doorkeeper-gem/doorkeeper/wiki/Changing-how-clients-are-authenticated
  # for more information on customization
  # access_token_methods :from_bearer_authorization, :from_access_token_param, :from_bearer_param

  # Change the native redirect uri for client apps
  # When clients register with the following redirect uri, they won't be redirected to any server and the authorization code will be displayed within the provider
  # The value can be any string. Use nil to disable this feature. When disabled, clients must provide a valid URL
  # (Similar behaviour: https://developers.google.com/accounts/docs/OAuth2InstalledApp#choosingredirecturi)
  #
  # native_redirect_uri 'urn:ietf:wg:oauth:2.0:oob'

  # Forces the usage of the HTTPS protocol in non-native redirect uris (enabled
  # by default in non-development environments). OAuth2 delegates security in
  # communication to the HTTPS protocol so it is wise to keep this enabled.
  #
  # Callable objects such as proc, lambda, block or any object that responds to
  # #call can be used in order to allow conditional checks (to allow non-SSL
  # redirects to localhost for example).
  #
  # force_ssl_in_redirect_uri !Rails.env.development?
  #
  # force_ssl_in_redirect_uri { |uri| uri.host != 'localhost' }

  # Specify what redirect URI's you want to block during creation. Any redirect
  # URI is whitelisted by default.
  #
  # You can use this option in order to forbid URI's with 'javascript' scheme
  # for example.
  #
  # forbid_redirect_uri { |uri| uri.scheme.to_s.downcase == 'javascript' }

  # Specify what grant flows are enabled in array of Strings. The valid
  # strings and the flows they enable are:
  #
  # "authorization_code" => Authorization Code Grant Flow
  # "implicit"           => Implicit Grant Flow
  # "password"           => Resource Owner Password Credentials Grant Flow
  # "client_credentials" => Client Credentials Grant Flow
  #
  # If not specified, Doorkeeper enables authorization_code and
  # client_credentials.
  #
  # implicit and password grant flows have risks that you should understand
  # before enabling:
  #   http://tools.ietf.org/html/rfc6819#section-4.4.2
  #   http://tools.ietf.org/html/rfc6819#section-4.4.3
  #
  # grant_flows %w[authorization_code client_credentials]

  # Hook into the strategies' request & response life-cycle in case your
  # application needs advanced customization or logging:
  #
  # before_successful_strategy_response do |request|
  #   puts "BEFORE HOOK FIRED! #{request}"
  # end
  #
  # after_successful_strategy_response do |request, response|
  #   puts "AFTER HOOK FIRED! #{request}, #{response}"
  # end

  # Under some circumstances you might want to have applications auto-approved,
  # so that the user skips the authorization step.
  # For example if dealing with a trusted application.
  # skip_authorization do |resource_owner, client|
  #   client.superapp? or resource_owner.admin?
  # end

  # WWW-Authenticate Realm (default "Doorkeeper").
  # realm "Doorkeeper"
end

#Doorkeeper.configuration.token_grant_types << "password"
```

Result is same - freeze

**Comment by nbulaj** [comment]
But if I run it in parallel, than yes, I see freeze:

> * Listening on tcp://localhost:9090
> Use Ctrl-C to stop
> Started POST "/oauth/token" for 127.0.0.1 at 2018-08-14 12:25:33 +0300
> Started POST "/oauth/token" for 127.0.0.1 at 2018-08-14 12:25:33 +0300
> Started POST "/oauth/token" for 127.0.0.1 at 2018-08-14 12:25:33 +0300
> Started POST "/oauth/token" for 127.0.0.1 at 2018-08-14 12:25:33 +0300
> 

:thinking:

**Comment by ViTVetal** [comment]
@nbulaj I got this problem when by mistake in Android app perform parallel two requests. So look like it isn't configuration mistake from my side?

**Comment by nbulaj** [comment]
No, @ViTVetal , it looks like dead-lock problem and requires some research

**Comment by nbulaj** [comment]
Yep, seems like dead-lock in [authorization_code_request.rb](https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/doorkeeper/oauth/authorization_code_request.rb#L31)

```ruby
ActiveRecord::StatementInvalid (SQLite3::BusyException: database is locked: UPDATE 
"oauth_access_grants" SET "revoked_at" = ? WHERE "oauth_access_grants"."id" = ?):
```

**Comment by ViTVetal** [comment]
@nbulaj Will you fix it? What should I do? Avoid situation with parallel requests?

**Comment by nbulaj** [comment]
@ViTVetal currently I can't say how to quickly fix this, because in normal situation we should not have this situation. The best we can is to handle transactions and locks. 

So from your side you now need to avoid situation with parallel requests, yep. From our side - fix the problem. Would you like to propose a PR with the fix?

**Comment by ViTVetal** [comment]
@nbulaj Thank you!
Sorry, I'm new to Rails so I will not risk propose a PR with the fix:-)

**Comment by bpieck** [comment]
I will try to reproduce and add a fix

**Comment by nbulaj** [comment]
@bpieck I was thinking about application level mutex for such operations to avoid transactions dead-locks, because splitting of mentioned above transaction is impossible

**Comment by bpieck** [comment]
Thans @nbulaj for additional directions :)

Sorry to reensure @nbulaj pointed into right direction ...

@ViTVetal did I understand correctly, that your mobile apps are doing a resource owner credentials grant - and no authorization code grant? I totally like to look into the authorization code grant, too. But I think it won't fix the issue @ViTVetal encountered?

**Comment by nbulaj** [comment]
Take a look at my curl example requests above, they are using auth code grant. I pointed the part of the code that is respobsible for the dead lock (same parallel requests that creates two transactions where one creates a lock and second raises with ActiveRecord::StatementError because of dead-lock)

**Comment by bpieck** [comment]
Yeah, I saw, what you did. But on @ViTVetal I just saw, that the request was sent to /oauth/token - and params where missing. And since he had shown his configuration:
```
grant_flows %w(password)
```
I assumed, he could only have made a resource owner credentials grant. But I assume you will be right and I was misreading :)

**Comment by nbulaj** [comment]
Hi @ViTVetal  @bpieck . I was trying to use application level mutex, but now I'm thinking about multiple app processes / servers and don't sure it is appropriate solution. Maybe we need to make token request faster / atomic / smth else to protect against dead-locks

**Comment by bpieck** [comment]
I am so sorry. Last I tried to reproduce the issue in a rspec test firstly to ensure I really find a fix for it - but failed. It's not to easy to simulate parallel request from within ruby ^^. So after that I lost track.

@ViTVetal What do you mean with: "I know that code is ugly and some parts not used". You just wrote wrong code examples and @nbulaj is right, that you did the auth code grant? Or that you did the resource owner credentials grant I assumed? In last case it would mean, nbulaj did not reproduce your issue, but found another (that maybe even not very probable really happens - I never ran into this issue on my side but encountered some mobile library, that did the request several times at once without running into this issue). Maybe you could write your example token request including params?

**Comment by ViTVetal** [comment]
@nbulaj Why this example is wrong? I wrote code with which the error occurred. 

about params: 
` curl -d "grant_type=refresh_token&refresh_token=ae373cf27eb9c4899aeecec463f6bebb10b00d22247b2a5c2e48b4d987e4ccbd" -X POST http://localhost:3000/oauth/token
`

If you will try perform this request parallel few times - you will get result described above

**Comment by bpieck** [comment]
So, you did a refresh_token request - not a authentication code grant. Thanks for pointing to it. @nbulaj mentioned something he reproduced on authentication code grant.

**Comment by bpieck** [comment]
I did not mean, that you wrote wrong code. I just wanted to understand what you meant with "I know that code is ugly and some parts not used".
If nbulaj's assumption about your issue was right (that your issue is rooted in https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/doorkeeper/oauth/authorization_code_request.rb#L31 ), then you would have done an authorization code grant and the configuration you posted:
```
  grant_flows %w(password)
```
Could not have matched your use case. So I wanted to know if "I know that code is ugly and some parts not used" means, you posted more then you really used. Just to ensure you really did the authorization code grant.
Still similar code exists on refresh token:
https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/doorkeeper/oauth/refresh_token_request.rb#L34
So most probably nbulaj is still just right with his assumption about the root cause. I maybe found some possible fix for it - but I was feeling very unsure, since I was not able to reproduce it with automatic tests. But I just wrote a possibly working test for it.

@nbulaj I think you are right, that the application level mutex will not solve the issue on a multi-server scenario. Give me two days to propose something. I am afraid my solution is not beautiful enough. But maybe you could go with it.

**Comment by nbulaj** [comment]
Made a small snippet (single-file Rails application with Doorkeeper) to improve the speed of tests for this issue :)

https://gist.github.com/nbulaj/f554f2c3382c3831b964e11afa5479d9

**Comment by nbulaj** [comment]
Seems like it's hard to reproduce such case IRL, so I'm closing this stale issue. But feel free to propose a PR or communicate over this issue.
